### PR TITLE
Feature/issue 183 persist more data

### PR DESCRIPTION
**Pull request by rasmus-storjohann-PG**

### CODE REVIEW

**Comment by rasmus-storjohann-PG** [comment]
@bensinjin I would suggest that you review this PR by looking at the diff for all the commits except the last one. The last commit touches a lot of code that the rest of the PR doesn't, so it's better to review the last commit separately. I just created a tmp branch at the second-to-last commit and diffed against master.

**Comment by rasmus-storjohann-PG** [comment]
Never mind, I removed the last commit from this PR, I'll put it up separately after this is merged.

File: src/sagas/questionnaire.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        console.error(`Failed to load active questions (${error.message})`);
+        yield put(Persistence.loadFailure(error.message));
+    }
+}
+
+export async function loadActiveQuestionsAsync(): Promise<string> {
+    return await AsyncStorage.getItem(ACTIVE_QUESTIONS_STORAGE_KEY);
+}
+
+const deserialize = (ids: string): ReadonlyArray<Id> | undefined => (
```
Comment (rasmus-storjohann-PG) [inline src/sagas/questionnaire.ts]: Remove undefined from return type

File: src/sagas/__tests__/questionnaire.test.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    it('should dispatch select effect to get active question ids from the store', () => {
+        const saga = saveActiveQuestions();
+
+        const result = saga.next().value;
+
+        expect(result).toEqual(select(selectIdsOfActiveQuestions));
+    });
+
+    describe('after selectig active question ids from store', () => {
```
Comment (bensinjin) [inline src/sagas/__tests__/questionnaire.test.ts]: Fix selectig typo

File: src/sagas/questionnaire.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+// tslint:disable:no-expression-statement
+
+import { call, CallEffect, PutEffect, put, ForkEffect, takeLatest, select, SelectEffect } from 'redux-saga/effects';
+import { AsyncStorage } from 'react-native';
+import { ACTIVE_QUESTIONS_STORAGE_KEY } from '../application/constants';
+import { Persistence, Id } from '../stores/questionnaire';
+import * as constants from '../application/constants';
+import { selectIdsOfActiveQuestions } from '../selectors/select_ids_of_active_questions';
```
Comment (bensinjin) [inline src/sagas/questionnaire.ts]: This name isn't great to me it looks as if it's going to return an array of question ids.  The term "Active" is somewhat vague I'd probably opt for something like "selected" or "chosen". For this function `selectChosenAnswerIds` or `getSelectedAnswerIds` would likely be an improvement.

File: src/sagas/questionnaire.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+export function* watchAnswerChangesToSaveAllAnswerStates(): IterableIterator<ForkEffect> {
+    yield takeLatest(constants.SELECT_ANSWER, saveActiveQuestions);
+}
+
+type SaveActions = IterableIterator<SelectEffect | CallEffect | PutEffect<Persistence.SaveSuccessAction | Persistence.SaveFailureAction>>;
+
+export function* saveActiveQuestions(): SaveActions {
+    try {
+        const ids = yield select(selectIdsOfActiveQuestions);
+        const serializedIds = serialize(ids);
```
Comment (bensinjin) [inline src/sagas/questionnaire.ts]: Is serialization a requirement here?

File: src/sagas/questionnaire.ts (hunk 3 above)
Comment (rasmus-storjohann-PG) [reply src/sagas/questionnaire.ts]: Yeah, I have a commit waiting that changes all instances of "selected" and "active" to "chosen", it was a mess.

Comment (bensinjin) [reply src/sagas/questionnaire.ts]: Also for my knowledge. Did you have a conventions around naming selectors? It seems we follow select<some-format><some-context> or select<some-context><some-format> did you have a preference?

Comment (rasmus-storjohann-PG) [reply src/sagas/questionnaire.ts]: That's all summarized in https://github.com/pg-irc/pathways-frontend/issues/123, please have a look and let me know if you disagree with any of it.

**Pull request merged**